frontends/common/constantFolding.cpp

```diff
@@ -237,6 +237,36 @@ const IR::Node *DoConstantFolding::preorder(IR::ArrayIndex *e) {
     return e;
 }
 
+// Returns true if the given expression is of the form "some_table.apply().action_run."
+bool isActionRun(const IR::Expression *e, const ReferenceMap *refMap) {
+    const auto *actionRunMem = e->to<IR::Member>();
+    if (!actionRunMem) return false;
+    if (actionRunMem->member.name != IR::Type_Table::action_run) return false;
+
+    const auto *applyMce = actionRunMem->expr->to<IR::MethodCallExpression>();
+    if (!applyMce) return false;
+    const auto *applyMceMem = applyMce->method->to<IR::Member>();
+    if (!applyMceMem) return false;
+    if (applyMceMem->member.name != IR::IApply::applyMethodName) return false;
+
+    const auto *tablePathExpr = applyMceMem->expr->to<IR::PathExpression>();
+    if (!tablePathExpr) return false;
+    const auto *tableDecl = refMap->getDeclaration(tablePathExpr->path);
+    if (!tableDecl) return false;
+
+    return tableDecl->is<IR::P4Table>();
+}
+
+const IR::Node *DoConstantFolding::preorder(IR::SwitchCase *c) {
+    // Action enum switch case labels must be action names.
+    // We only want to fold the label expression after it is inspected by the typechecker.
+    const auto *parent = static_cast<const IR::SwitchStatement *>(getContext()->node);
```

Comment:
Use `->to<IR::SwitchStatement>()`.

Reply:
I did this to avoid a dynamic cast, as a `SwitchCase`'s parent is always `SwitchStatement`

Reply:
How about doing the preorder directly on the SwitchStatement then? Otherwise, I would just add a comment explaining why the static_cast was used.

Reply:
I gave a comment on the link. I don't think it is possible to directly visit a node in a transform visitor. You need to consume the result of the `apply` call with an assignment instead. In this case I suspect that one needs to assign a new list of switch cases to the `IR::SwitchStatement` node.